tests/test_fields.py

```diff
@@ -1,6 +1,392 @@
+    )
+    assert (
+        len(lognormal_gls([np.linspace(1, 5, 5), np.linspace(1, 5, 5)], shift)[1]) == 5
+    )
+
+
+def test_discretized_cls() -> None:
+    # empty cls
+
+    result = discretized_cls([])
+    assert result == []
+
+    # power spectra truncated at lmax + 1 if lmax provided
+
+    result = discretized_cls([np.arange(10), np.arange(10), np.arange(10)], lmax=5)
+
+    for cl in result:
+        assert len(cl) == 6
+
+    # check ValueError for triangle number
+
+    with pytest.raises(
+        ValueError, match="length of cls array is not a triangle number"
+    ):
+        discretized_cls([np.arange(10), np.arange(10)], ncorr=1)
+
+    # ncorr not None
+
+    cls = [np.arange(10), np.arange(10), np.arange(10)]
+    ncorr = 0
+    result = discretized_cls(cls, ncorr=ncorr)
+
+    assert len(result[0]) == 10
+    assert len(result[1]) == 10
+    assert len(result[2]) == 0  # third correlation should be removed
+
+    # check if pixel window function was applied correctly with nside not None
+
+    nside = 4
+
+    pw = hp.pixwin(nside, lmax=7)
+
+    result = discretized_cls([[], np.ones(10), np.ones(10)], nside=nside)
+
+    for cl in result:
+        n = min(len(cl), len(pw))
+        expected = np.ones(n) * pw[:n] ** 2
+        np.testing.assert_allclose(cl[:n], expected)
+
+
+def test_effective_cls() -> None:
+    # empty cls
+
+    result = effective_cls([], np.array([]))
+    assert result.shape == (0,)
+
+    # check ValueError for triangle number
+
+    with pytest.raises(ValueError, match="length of cls is not a triangle number"):
+        effective_cls([np.arange(10), np.arange(10)], np.ones((2, 1)))
+
+    # check ValueError for triangle number
+
+    with pytest.raises(ValueError, match="shape mismatch between fields and weights1"):
+        effective_cls([], np.ones((3, 1)))
+
+    # check with only weights1
+
+    cls = [np.arange(15), np.arange(15), np.arange(15)]
+    weights1 = np.ones((2, 1))
+
+    result = effective_cls(cls, weights1)
+    assert result.shape == (1, 1, 15)
+
+    # check truncation if lmax provided
+
+    result = effective_cls(cls, weights1, lmax=5)
+
+    assert result.shape == (1, 1, 6)
+    np.testing.assert_allclose(result[..., 6:], 0)
+
+    # check with weights1 and weights2 and weights1 is weights2
+
+    result = effective_cls(cls, weights1, weights2=weights1)
+    assert result.shape == (1, 1, 15)
+
+
+def test_generate_gaussian() -> None:
+    gls = [np.array([1.0, 0.5, 0.1])]
+    nside = 4
+    ncorr = 1
+
+    gaussian_fields = list(generate_gaussian(gls, nside))
+
+    assert gaussian_fields[0].shape == (hp.nside2npix(nside),)
+
+    # requires resetting the RNG for reproducibility
+    rng = np.random.default_rng(seed=42)
+    gaussian_fields = list(generate_gaussian(gls, nside, rng=rng))
+
+    assert gaussian_fields[0].shape == (hp.nside2npix(nside),)
+
+    # requires resetting the RNG for reproducibility
+    rng = np.random.default_rng(seed=42)
```

Comment:
Why have it twice within this function?

Reply:
Should we set it once at the top of the file maybe?

Reply:
Need it twice to ensure that `rng.any_distribution()` is same both the times, and it needs to be redefined (instead of using `rng` from the arguments) so that it starts sampling from scratch.

Setting it at the top of the file gives me -
```py
        # requires resetting the RNG for reproducibility
>       gaussian_fields = list(generate_gaussian(gls, nside, rng=rng))
E       UnboundLocalError: cannot access local variable 'rng' where it is not associated with a value
```

Reply:
Okay! Might a little comment saying why it's being redefined?

Reply:
I think the right approach here is to use our central `rng` fixture and clone it:

```py
from copy import deepcopy

def test_xyz(rng):
    rng1 = deepcopy(rng)
    rng2 = deepcopy(rng)
    ...
```

Reply:
This probably needs a fairly recent numpy

Reply:
But what is this actually testing? If it's that the generation is reproducible, it's on the RNG, not on us. In which case I would rather test that the rng was actually used:

```py
initial_rng_state = rng.bit_generator.state

...

assert rng.bit_generator.state != initial_rng_state
```

Reply:
My initial thought behind adding this test was to check if `generate_gaussian` with and without `ncorr=1` outputs the same result (because that is a pattern that I observed while running the function manually). Is there a better test that can be added to check if passing the `ncorr` argument works, or should I leave it for #414?